File: KisaragiTech.Dape/User/Database/UserRepositoryImpl.cs
using System;
using System.Threading.Tasks;
using KisaragiTech.Dape.User.Interface;
using KisaragiTech.Dape.User.Model;
using Neo4j.Driver;

namespace KisaragiTech.Dape.User.Database;

internal sealed class UserRepositoryImpl
{
    private readonly IDriver driver;
    
    internal UserRepositoryImpl(IDriver driver)
    {
        this.driver = driver;
    }
    
    public async Task<bool> HasRootUser()
    {
        await using var session = driver.AsyncSession(o => o.WithDefaultAccessMode(AccessMode.Read));
        
        return await session.ExecuteReadAsync(async tx =>
        {
            // language=cypher
            const string query = "MATCH (u:User { root: true }) RETURN u LIMIT 1";
            var cursor = await tx.RunAsync(query);
            return await cursor.FetchAsync();
        });
    }

    public async Task<bool> FindUserByPreferredHandle(string preferredHandle)
    {
        await using var session = driver.AsyncSession(sess => sess.WithDefaultAccessMode(AccessMode.Read));

        return await session.ExecuteReadAsync(async tx =>
        {
            // language=cypher
            const string query = "MATCH (n:Person { handle: $handle }) RETURN true AS exists LIMIT 1";
            var cursor = await tx.RunAsync(query, new { handle = preferredHandle });

            return await cursor.FetchAsync();
        });
    }

    private async Task AssertPreferredHandleUniqueness(IUser user)
    {
        if (await this.FindUserByPreferredHandle(user.GetPreferredHandle()))
        {
            throw new ArgumentException($"すでに {user.GetPreferredHandle()} は登録されています。");
        }
    }

    public async Task InsertUser(LocalRegisteredUser user)
    {
        await this.AssertPreferredHandleUniqueness(user);
        
        await using var session = driver.AsyncSession(sess => sess.WithDefaultAccessMode(AccessMode.Write));

        await session.ExecuteWriteAsync(tx =>
        {
            // language=cypher
            const string query = "CREATE (n:Person { handle: $handle, id: $id })";
            var parameters = new { handle = user.GetPreferredHandle(), id = user.GetIdentifier().Raw.ToString() };

            return tx.RunAsync(query, parameters);
        });
    }

    public async Task CreateRootUser(LocalRegisteredUser user)
    {
        if (await this.HasRootUser())
        {
            throw new InvalidOperationException("ルートユーザーはすでに存在します");
        }

        await InsertUser(user);
    }
}

File: src/main/scala/io/github/ksrgtech/dape/user/database/UserRepositoryImpl.scala
package io.github.ksrgtech.dape.user.database

import cats.effect.kernel.Async
import cats.implicits.*
import neotypes.AsyncDriver
import neotypes.mappers.ResultMapper
import neotypes.syntax.all.*
import io.github.ksrgtech.dape.user.model.{LocalRegisteredUser, User}
import io.github.ksrgtech.dape.base.cats.assert

final class UserRepositoryImpl[F[_]: Async](driver: AsyncDriver[F]) extends UserRepository[F] {

  override def hasRootUser: F[Boolean] = {
    val query = "MATCH (u:User { root: true }) RETURN u LIMIT 1"
    query.query(ResultMapper.option(using ResultMapper.ignore)).single(driver).map(_.isDefined)
  }

  override def findUserByPreferredHandle(preferredHandle: String): F[Boolean] =
    c"MATCH (n:User { handle: $preferredHandle }) RETURN true AS exists LIMIT 1"
      .query(ResultMapper.option(using ResultMapper.boolean))
      .single(driver)
      .map(_.getOrElse(false))

  private def assertPreferredHandleUniqueness(user: User): F[Unit] =
    findUserByPreferredHandle(user.preferredHandle).flatMap { exists =>
      if exists then {
        Async[F].raiseError(
          new IllegalArgumentException(s"Already registered: ${user.preferredHandle}")
        )
      } else {
        Async[F].unit
      }
    }

  override def insertUser(user: LocalRegisteredUser): F[Unit] =
    // FIXME: TOCTOU
    for {
      _ <- assertPreferredHandleUniqueness(user)
      handle = user.preferredHandle
      id     = user.getIdentifier.raw.toString
      _ <- c"CREATE (n:User { handle: $handle, id: $id })".execute.void(driver)
    } yield ()

  override def createRootUser(user: LocalRegisteredUser): F[Unit] =
    // FIXME: TOCTOU
    for {
      _        <- assertPreferredHandleUniqueness(user)
      haveRoot <- this.hasRootUser
      _        <- Async[F].assert(haveRoot, "root user has been created!")
      handle = user.preferredHandle
      id     = user.getIdentifier.raw.toString
      _ <- c"CREATE (n:User:Person { handle: $handle, id: $id, root: true })".execute.void(driver)
    } yield ()
}
